File: packages/patapata_core/android/src/main/kotlin/dev/patapata/patapata_core/PatapataCorePlugin.kt
/*
 * Copyright (c) GREE, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package dev.patapata.patapata_core

import android.util.Log
import androidx.annotation.NonNull
import androidx.annotation.Keep
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor
import io.flutter.embedding.engine.loader.FlutterLoader

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File

private class PatapataPluginContainer(val plugin: PatapataPlugin, val flutterPluginBinding: FlutterPlugin.FlutterPluginBinding)

private val sPlugins = mutableSetOf<PatapataPluginContainer>()

/** PatapataCorePlugin */
class PatapataCorePlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var mChannel : MethodChannel
  private lateinit var mBinding : FlutterPlugin.FlutterPluginBinding

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    mBinding = flutterPluginBinding
    // Lines up with Flutter's override of Dart's Directory.systemTemp directory.
    val tNativeLibDirectoryInfoFile = File(flutterPluginBinding.applicationContext.codeCacheDir, "patapataNativeLib")

    if (tNativeLibDirectoryInfoFile.exists()) {
      tNativeLibDirectoryInfoFile.delete()
    }

    tNativeLibDirectoryInfoFile.writeText(flutterPluginBinding.applicationContext.applicationInfo.nativeLibraryDir)

    mChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "dev.patapata.patapata_core")
    mChannel.setMethodCallHandler(this)

    // Register default plugins.
    flutterPluginBinding.registerPatapataPlugin(NativeLocalConfig(mBinding.applicationContext, mBinding.binaryMessenger))
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "enablePlugin" -> {
        val tName = call.arguments as? String

        if (tName == null) {
          result.error(Error.PPE000.name, "Invalid plugin name passed to enablePlugin", null)

          return
        }

        enablePlugin(tName)
        result.success(null)
      }
      "disablePlugin" -> {
        val tName = call.arguments as? String

        if (tName == null) {
          result.error(Error.PPE000.name, "Invalid plugin name passed to disablePlugin", null)

          return
        }

        disablePlugin(tName)
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    mChannel.setMethodCallHandler(null)
  }

  private fun enablePlugin(pluginName: String) {
    sPlugins.firstOrNull {
      it.plugin.patapataName == pluginName && it.flutterPluginBinding.binaryMessenger == mBinding.binaryMessenger
    }?.plugin?.patapataEnable()
  }

  private fun disablePlugin(pluginName: String) {
    sPlugins.firstOrNull {
      it.plugin.patapataName == pluginName && it.flutterPluginBinding.binaryMessenger == mBinding.binaryMessenger
    }?.plugin?.patapataDisable()
  }
}

fun FlutterPlugin.FlutterPluginBinding.registerPatapataPlugin(plugin: PatapataPlugin) {
  sPlugins.removeAll { it.plugin == plugin }
  sPlugins.add(PatapataPluginContainer(plugin, this))
}

fun FlutterPlugin.FlutterPluginBinding.unregisterPatapataPlugin(plugin: PatapataPlugin) {
  sPlugins.removeAll { it.plugin == plugin }
}

File: packages/patapata_core/ios/Classes/PatapataCorePlugin.swift
// Copyright (c) GREE, Inc.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

import Flutter
import UIKit
import AppTrackingTransparency

fileprivate var sPlugins = Set<PatapataPluginContainer>()

public class PatapataCorePlugin: NSObject, FlutterPlugin {
    let mRegistrar: FlutterPluginRegistrar
    let mChannel: FlutterMethodChannel
    
    init(registrar: FlutterPluginRegistrar) {
        mRegistrar = registrar
        mChannel = FlutterMethodChannel(name: "dev.patapata.patapata_core", binaryMessenger: registrar.messenger())
        
        super.init()
        
        registrar.addMethodCallDelegate(self, channel: mChannel)
        
        // Register default plugins.
        registrar.registerPatapata(plugin: NativeLocalConfig(registrar: registrar))
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let _ = PatapataCorePlugin(registrar: registrar)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "enablePlugin":
            guard let tName = call.arguments as? String else {
                result(FlutterError(code: "PPE000", message: "Invalid plugin name passed to enablePlugin", details: nil))
                
                return
            }
            
            enablePlugin(with: tName)
            result(nil)
        case "disablePlugin":
            guard let tName = call.arguments as? String else {
                result(FlutterError(code: "PPE000", message: "Invalid plugin name passed to disablePlugin", details: nil))
                
                return
            }
            
            disablePlugin(with: tName)
            result(nil)
        case "Permissions:requestTracking":
            requestTrackingPermission(result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    public func enablePlugin(with pluginName: String) {
        for i in sPlugins {
            guard i.plugin.patapataName == pluginName && i.registrar.messenger().hash == mRegistrar.messenger().hash else {
                continue
            }
            
            i.plugin.patapataEnable();
        }
    }
    
    public func disablePlugin(with pluginName: String) {
        for i in sPlugins {
            guard i.plugin.patapataName == pluginName && i.registrar.messenger().hash == mRegistrar.messenger().hash else {
                continue
            }
            
            i.plugin.patapataDisable()
        }
    }
    
    func requestTrackingPermission(_ result: @escaping FlutterResult) {
        if #available(iOS 14, *) {
            ATTrackingManager.requestTrackingAuthorization { (status) in
                switch status {
                case .authorized:
                    result("authorized")
                    break
                case .denied:
                    result("denied")
                    break
                case .notDetermined:
                    result("notDetermined")
                    break
                case .restricted:
                    result("restricted")
                    break
                default:
                    result(nil)
                }
            }
        } else {
            result(nil)
        }
    }
}

fileprivate struct PatapataPluginContainer : Hashable {
    let plugin: PatapataPlugin
    let registrar: FlutterPluginRegistrar
    
    static func == (lhs: PatapataPluginContainer, rhs: PatapataPluginContainer) -> Bool {
        return lhs.plugin.patapataName == rhs.plugin.patapataName && lhs.registrar.messenger().hash == rhs.registrar.messenger().hash
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(plugin.patapataName)
        hasher.combine(registrar.messenger().hash)
    }
}


extension FlutterPluginRegistrar {
    public func registerPatapata(plugin: PatapataPlugin) {
        sPlugins.insert(PatapataPluginContainer(plugin: plugin, registrar: self))
    }
    
    public func unregisterPatapata(plugin: PatapataPlugin) {
        sPlugins.remove(PatapataPluginContainer(plugin: plugin, registrar: self))
    }
}
